oauth2client/contrib/multiprocess_file_storage.py

```diff
@@ -0,0 +1,261 @@
+            return
+
+        self._credentials.update(self._load_credentials_file(self._file))
+
+        logger.debug('Read credential file')
+
+    def _load_credentials_file(self, fh):
+        credentials = {}
+
+        try:
+            fh.seek(0)
+            data = json.load(fh)
+        except Exception:
+            logger.warning(
+                'Credentials file could not be loaded, will ignore and '
+                'overwrite.')
+            return credentials
+
+        if data.get('file_version') != 2:
+            data = {}
+            logger.warning(
+                'Credentials file is not version 2, will ignore and '
+                'overwrite.')
+            return credentials
+
+        for key, encoded_credential in iteritems(data.get('credentials', {})):
+            try:
+                credential_json = base64.b64decode(encoded_credential)
+                credential = Credentials.new_from_json(credential_json)
+                credentials[key] = credential
+            except:
+                logger.warning(
+                    'Invalid credential {} in file, ignoring.'.format(key))
+
+        return credentials
+
+    def _write_credentials(self):
+        if self._read_only:
+            logger.debug('In read-only mode, not writing credentials.')
+            return
+
+        self._create_file_if_needed()
+        self._write_credentials_file(self._file, self._credentials)
+        logger.debug('Wrote credential file {}.'.format(self._filename))
+
+    def _write_credentials_file(self, fh, credentials):
+        data = {'file_version': 2, 'credentials': {}}
+
+        for key, credential in iteritems(credentials):
+            credential_json = credential.to_json()
+            encoded_credential = base64.b64encode(credential_json)
+            data['credentials'][key] = encoded_credential
+
+        fh.seek(0)
+        json.dump(data, fh)
+        fh.truncate()
+
+    def acquire_lock(self):
+        self._thread_lock.acquire()
+        locked = self._process_lock.acquire(timeout=INTERPROCESS_LOCK_DEADLINE)
+
+        self._create_file_if_needed()
+
+        if locked:
+            self._file = open(self._filename, 'r+')
+            self._read_only = False
+
+        else:
+            self._read_only = True
+            logger.warn(
+                'Failed to obtain interprocess lock for credentials. '
+                'If a credential is being refreshed, other processes may '
+                'not see the updated access token and refresh as well.')
+            self._file = open(self._filename, 'r')
+
+        self._load_credentials()
+
+    def release_lock(self):
+        if self._file is not None:
+            self._file.close()
+            self._file = None
+
+        if not self._read_only:
+            self._process_lock.release()
+
+        self._thread_lock.release()
+
+    def _refresh_predicate(self, credentials):
+        if credentials is None:
```

Comment:
I thought I felt strongly... but nowhere near strongly enough to take on the both of you. Carry on for now; I reserve the right to come back in the future with a reasonably persuasive argument if I can find one.